Bluetooth: tester: Increase the adv buf size to consider the extend adv

nngt88 opened the pull request:
[Description]
Device hang is observed when LE Scan
[Root Cause]
For le_ext_adv_report, Data[i] of one adv report is 0-229,
for the adv_buf of saving, it just is 73 bytes, so sometimes
the bytes of adv report is more than the adv_buf, resulted hang
[Fix]
consider the le ext adv case.
[Testing]
After modified, Device hang is not observed after stress
testing with LE Scan

Signed-off-by: Guotao Zhang <guotao.zhang@nxp.com>

sjanc commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -30,7 +30,8 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
 #define BT_LE_AD_DISCOV_MASK (BT_LE_AD_LIMITED | BT_LE_AD_GENERAL)
-#define ADV_BUF_LEN (sizeof(struct btp_gap_device_found_ev) + 2 * 31)
+/*consider ext_adv case*/
+#define ADV_BUF_LEN (sizeof(struct btp_gap_device_found_ev) + 2 * 229)
```
Hmm I don't think we need  2*229 as for ext adv there is either AD or SCAN_RSP, never both.
Also, maybe this should be CONFIG_BT_EXT_SCAN_BUF_SIZE instead of hardcoded 229 ?


And this buffer is used to concatenate AD and SCAN_RSP in BTP event.

nngt88 replied:
Hi @sjanc   The macro CONFIG_BT_EXT_ADV is set yes in bt_tester project,  so when start scan, actually start le ext scan, 
the data len of scan rsp maybe  more than 31.

sjanc replied:
Hi,

I mean that this could be
#define ADV_BUF_LEN (sizeof(struct btp_gap_device_found_ev) + CONFIG_BT_EXT_SCAN_BUF_SIZE);

or maybe to be sure we always fit full legacy AD+SCAN_RSP
#define ADV_BUF_LEN (sizeof(struct btp_gap_device_found_ev) + MAX(2 * 31, CONFIG_BT_EXT_SCAN_BUF_SIZE));

It should cover both cases since in ext adv PDUs it is either AD or SCAN_RSP, never both

nngt88 replied:
Hi @sjanc Thanks your comment, I have modified the code, can you help review again

nngt88 replied:
Hi @sjanc  Do you have any other suggestions on your end

nngt88 committed:
Bluetooth: tester: Increase the adv buf size to consider the extend adv

[Description]
Device hang is observed when LE Scan
[Root Cause]
For le_ext_adv_report, Data[i] of one adv report is 0-229,
for the adv_buf of saving, it just is 73 bytes, so sometimes
the bytes of adv report is more than the adv_buf, resulted hang
[Fix]
consider the le ext adv case.
[Testing]
After modified, Device hang is not observed after stress
testing with LE Scan

Signed-off-by: Guotao Zhang <guotao.zhang@nxp.com>

nngt88 commented:
Hi @everyone     Can you help review the code. Thanks

alwa-nordic commented:
I suggest enabling asserts in the tester project. It enables bounds-checking in net_buf functions. Otherwise, I suggest adding bounds checks to `btp_gap.c` where `adv_buf` is written to.

alwa-nordic commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -30,7 +30,11 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
 
 #define BT_LE_AD_DISCOV_MASK (BT_LE_AD_LIMITED | BT_LE_AD_GENERAL)
+#if defined(CONFIG_BT_EXT_ADV)
+#define ADV_BUF_LEN (sizeof(struct btp_gap_device_found_ev) + CONFIG_BT_EXT_SCAN_BUF_SIZE)
```
I think the report can be larger than `CONFIG_BT_EXT_SCAN_BUF_SIZE`. When the report is not fragmented, the Host uses the driver buffer directly, so the report can be up to `BT_BUF_EVT_DISCARDABLE_SIZE` minus some number of header bytes.

On this code path:

https://github.com/zephyrproject-rtos/zephyr/blob/f2b6490dee30b38dfb0ee31902177491c35ebacb/subsys/bluetooth/host/scan.c#L659

nngt88 replied:
Hi @alwa-nordic    Thanks for your comment, I have modified the code to 2 *  CONFIG_BT_EXT_SCAN_BUF_SIZE, it should    
enough, can you help review again

nngt88 commented:
> I suggest enabling asserts in the tester project. It enables bounds-checking in net_buf functions. Otherwise, I suggest adding bounds checks to `btp_gap.c` where `adv_buf` is written to.

Hi Thanks your comment, the assert is enable in the tester project, I just got the assert when enable the le ext adv, because the space of adv_buf is not enough when add to adv_buff.
![image](https://github.com/user-attachments/assets/01055379-240d-44dc-8863-54a9a5d81df1)

alwa-nordic commented:
> > I suggest enabling asserts in the tester project. It enables bounds-checking in net_buf functions. Otherwise, I suggest adding bounds checks to `btp_gap.c` where `adv_buf` is written to.
>
> the assert is enable in the tester project, I just got the assert

Ok, thanks for confirming that. Since the [PR description](https://github.com/zephyrproject-rtos/zephyr/pull/75756#issue-2402469299) says "device hangs", I though it was a deadlock or lack or response of some kind, and the assert panic and crash was not happening.

I also checked `tester/prj.conf`, and there was no `CONFIG_ASSERT=y`. Though it looks like it's enabled for particular boards, e.g. in `boards/nrf52840dk_nrf52840.conf`. This may be something the maintainers of tester want to look into.

nngt88 commented:
> > > I suggest enabling asserts in the tester project. It enables bounds-checking in net_buf functions. Otherwise, I suggest adding bounds checks to `btp_gap.c` where `adv_buf` is written to.
> > 
> > 
> > the assert is enable in the tester project, I just got the assert
> 
> Ok, thanks for confirming that. Since the [PR description](https://github.com/zephyrproject-rtos/zephyr/pull/75756#issue-2402469299) says "device hangs", I though it was a deadlock or lack or response of some kind, and the assert panic/crash was not happening.
> 
> I also checked `tester/prj.conf`, and there was no `CONFIG_ASSERT=y`. Though it looks like it's enabled for particular boards, e.g. in `boards/nrf52840dk_nrf52840.conf`. This may be something the maintainers of tester want to look into.

Hi @alwa-nordic  Sorry later reply, I have checked the code again, the ASSERT default y  in all tests project, so I think we don't need to add the CONFIG_ASSERT=y in the tester/prj.conf. Do you think so?
This is the PR enable assertions for all tests. I checked it helpful for the tester project. https://github.com/nngt88/zephyr/commit/bc40f9340db2232830b8ac9eb5bd2e2c41dc9509

alwa-nordic commented:
> > > > I suggest enabling asserts in the tester project. It enables bounds-checking in net_buf functions. Otherwise, I suggest adding bounds checks to `btp_gap.c` where `adv_buf` is written to.
> > > 
> > > 
> > > the assert is enable in the tester project, I just got the assert
> > 
> > 
> > Ok, thanks for confirming that. Since the [PR description](https://github.com/zephyrproject-rtos/zephyr/pull/75756#issue-2402469299) says "device hangs", I though it was a deadlock or lack or response of some kind, and the assert panic/crash was not happening.
> > I also checked `tester/prj.conf`, and there was no `CONFIG_ASSERT=y`. Though it looks like it's enabled for particular boards, e.g. in `boards/nrf52840dk_nrf52840.conf`. This may be something the maintainers of tester want to look into.
> 
> Hi @alwa-nordic Sorry later reply, I have checked the code again, the ASSERT default y in all tests project, so I think we don't need to add the CONFIG_ASSERT=y in the tester/prj.conf. Do you think so? This is the PR enable assertions for all tests. I checked it helpful for the tester project. [nngt88@bc40f93](https://github.com/nngt88/zephyr/commit/bc40f9340db2232830b8ac9eb5bd2e2c41dc9509)

Ok, all good then. I was not aware `CONFIG_TEST` enables `CONFIG_ASSERT` before.
